Srpg 39  character base

chai227chai opened the pull request:
## 개요
<!---- 변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->

캐릭터 클래스와 스킬 클래스 새로 생성했습니다.
캐릭터 클래스의 경우 prefab으로 모든 캐릭터의 오브젝트를 미리 만들어 둔 후 
dictionary나 list를 통해 관리하는 식으로 만들고 싶어서 Monobehavior를 상속 시켜놨는데, 추후 설계 방향에 따라서 변경될 수 있습니다.

<!---- Resolves: #(Isuue Number) -->

## PR 유형
어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] CSS 등 사용자 UI 디자인 변경
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [ ] 코드 리팩토링
- [x] 주석 추가 및 수정
- [ ] 문서 수정
- [x] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다. (ex. feature/SRPG-11-UIBase)
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

chai227chai committed:
feat : 캐릭터 이동 기능 추가

chai227chai committed:
refactor : 캐릭터에 이동 반경 추가

chai227chai committed:
feat : CharacterSo, SkillSo 추가

chai227chai committed:
feat : Character클래스 추가

Character클래스를 언제 선언해서 사용할 지 고민 중

chai227chai committed:
feat : 스킬 범위 구현

chai227chai committed:
feat : 스킬 효과 패시브 효과 작업, 추석 추가

chai227chai commented on Assets/Scripts/Objects/Characters/CharacterInBattle/CharacterBase.cs:
```diff
@@ -0,0 +1,149 @@
+        curCharacterPassive?.OnStartAttack(enemy);
+    }
+
+    public virtual void OnAttackSuccess(CharacterBase enemy, int damage)// ���� ���� ��
+    {
+        curCharacterPassive?.OnAttackSuccess(enemy, damage);
+    }
+
+    public virtual void OnEndAttack(List<CharacterBase> enemy)// ���� ���� ��
+    {
+        curCharacterPassive?.OnEndAttack(enemy);
+    }
+
+    public virtual void OnTakeDamage(CharacterBase enemy)// ���� �޾��� ��
+    {
+        curCharacterPassive?.OnTakeDamage(enemy);
+    }
+
+    //---------------------------------------------------------------------------
+    // ��ų ����
+
+    public virtual void OnUseSkill(List<CharacterBase> target)// ��ų ��� �� 
+    {
+        curCharacterSkill.skillAbility?.OnUseSkill(target);
+    }
+
+    public virtual void OnSkillAttackSuccess(CharacterBase target, int damage)// ��ų ���� ���� ��
+    {
+        if (curCharacterSkill.skillData.onhit)
+        {
+            curCharacterPassive?.OnAttackSuccess(target, damage);
+        }
+        curCharacterSkill.skillAbility?.OnSkillAttackSuccess(target, damage);
+    }
+
+    public virtual void OnEndSkill(List<CharacterBase> target)// ��ų ��� ���� ��
+    {
+        curCharacterSkill.skillAbility?.OnEndSkill(target);
+    }
+}
```
이 부분에 굳이 행동 별 함수를 다시 선언한 이유는 같은 상황일 때 발동하는 효과들을 한꺼번에 관리하기 위해서 입니다

SeRDic0705 reviewed (commented):
너무 멋지게 작업해주셨는데요?
- 양이 너무 많아서 변경사항을 확인이 힘들어요(PR 조금씩 나눠서 자주하면 좋을 것 같아요)
- 주석이 있는 스크립트도 있고 없는 스크립트도 있는데, 없는곳에서는 어떤 로직인지 한눈에 알기가 어려워요

메서드 별로만이라도 주석을 추가해주실 수 있을까요?

GeCrabKani reviewed (changes requested):
MapManager에 중복된 코드가 있습니다.

    Vector2Int[] directions = new Vector2Int[]
    {
        new Vector2Int(1, 0), new Vector2Int(-1, 0),
        new Vector2Int(0, 1), new Vector2Int(0, -1),
        new Vector2Int(1, 1), new Vector2Int(-1, 1),
        new Vector2Int(1, -1), new Vector2Int(-1, -1),
    };

    foreach (Vector2Int dir in directions)
    {
        Vector2Int tileToCheck = originTile + dir;
        if (map.ContainsKey(tileToCheck))
        {
        }
    }
와 같이 방향 벡터 배열을 이용하면 중복 코드를 줄일 수 있습니다.




CharacterController에 PlayerPhase 타입이 클래스 밖에 public 타입으로 선언되어있습니다.
타 스크립트에서 사용이 필요하다면 Utils의 Constants에 선언해주세요.
해당 스크립트에서만 사용한다면 클래스 내부에 private로 선언해주세요.

CharacterSO의 Enum들도 해당





MapTiles의 Awake에서 매니저 Init()을 실행하고 있습니다.
Managers 스크립트의 get 부분에서 Init을 실행 후 return을 해주고 있습니다.
타 스크립트에서 사용 시 Init 순서를 신경쓰지 않아도 괜찮습니다.

chai227chai committed:
feat & refactor : 주석 추가 및 이동 매커니즘 변경

chai227chai committed:
feat & refactor : 주석 추가 및 코드 정리

chai227chai reviewed (commented):
피드백 부분을 수정했습니다.

SeRDic0705 commented on Assets/Scripts/MapScripts/CharacterController/CharacterController.cs:
```diff
@@ -197,7 +197,7 @@ private void GetPathTile()
 
     private void GetMoveRangeTile()// �̵� ���� �Ÿ� ������
```
주석들이 깨져서 나와요

인코딩 설정이 변경된 것 같은데 UTF-8인지 확인해서 수정 부탁드려요

chai227chai replied:
파일 변경 완료 했습니다

chai227chai committed:
feat : 파일 유니코드 변경

jhwoo944 committed:
SRPG-23 feat: 캐릭터 성장 능력치 추가 외

-CharacterSO에 GrowStatus 추가
-공격타입 Mage와 Res 능력치는 더 이상 사용하지 않으므로 삭제
-ApplyGrowStat(  ) 메서드 내용 추가
//원본캐릭터의 기본 스탯 + (원본 캐릭터의 성장 스탯 * 이 캐릭터 개체의 현재 레벨)
-Constants에 Character 관련 상수값 enum 추가 (현재 작성된 상수: 캐릭터 기본 치명타확률과 기본 치명타 피해)
